### PR TITLE
core/state: cache storage tries across blocks and pending state

### PR DESCRIPTION
**Pull request by karalabe**

### CODE REVIEW

**Comment by fjl** [comment]
The code looks OK to me, but I don't understand why the locked trie wrapper is needed. Can you explain that part?

File: core/state/database.go [hunk 1]
```diff
@@ -176,3 +239,151 @@ func (m cachedTrie) Commit(onleaf trie.LeafCallback) (common.Hash, error) {
 func (m cachedTrie) Prove(key []byte, fromLevel uint, proofDb ethdb.Putter) error {
 	return m.SecureTrie.Prove(key, fromLevel, proofDb)
 }
+
+// syncedTrie is a synchronized wrapper around a trie to support multiple threads
+// accessing and read-expanding the same trie (i.e. caching loaded trie nodes).
+type syncedTrie struct {
+	tr   *trie.SecureTrie
+	lock sync.Mutex
```
Comment (ryanschneider) [inline core/state/database.go]: Any reason not to use a `sync.RWMutex` here instead?  If there's a strong argument for it being a traditional mutex perhaps a comment should be added so others don't try to update it to a RWMutex later?

Comment (karalabe) [reply core/state/database.go]: A trie gets modified if a read expands a new path as it needs to load data from disk. As such, any operation on the trie needs to be exclusive.

**Comment by karalabe** [comment]
Closing in favor of https://github.com/ethereum/go-ethereum/pull/18087, which performs better, uses allowed memory to the full, but also has a guaranteed hard cap on memory use.

**Pull request closed without merging**